infrastructure/comms/src/peer_manager/node_id.rs

```diff
@@ -0,0 +1,252 @@
+    pub fn distance(&self, node_id: &NodeID) -> NodeDistance {
+        (xor_metric(&self, &node_id))
+    }
+
+    /// Find and return the indices of the K nearest neighbours from the provided node id list
+    pub fn closest_indices(&self, node_ids: &Vec<NodeID>, k: usize) -> Result<Vec<usize>, NodeIDError> {
+        if k > node_ids.len() {
+            return Err(NodeIDError::OutOfBounds);
+        }
+        let mut indices: Vec<usize> = Vec::with_capacity(node_ids.len());
+        let mut dists: Vec<NodeDistance> = Vec::with_capacity(node_ids.len());
+        for i in 0..node_ids.len() {
+            indices.push(i);
+            dists.push(self.distance(&node_ids[i]));
+        }
+        // Perform partial sort of elements only up to K elements
+        let mut nearest_node_indices: Vec<usize> = Vec::with_capacity(k);
+        for i in 0..k {
+            for j in i + 1..node_ids.len() {
+                if dists[i] > dists[j] {
+                    dists.swap(i, j);
+                    indices.swap(i, j);
+                }
+            }
+            nearest_node_indices.push(indices[i]);
+        }
+        Ok(nearest_node_indices)
+    }
+
+    /// Find and return the node ids of the K nearest neighbours from the provided node id list
+    pub fn closest(&self, node_ids: &Vec<NodeID>, k: usize) -> Result<Vec<NodeID>, NodeIDError> {
+        match self.closest_indices(&node_ids, k) {
+            Ok(nearest_node_indices) => {
+                let mut nearest_node_ids: Vec<NodeID> = Vec::with_capacity(nearest_node_indices.len());
+                for i in 0..nearest_node_indices.len() {
+                    nearest_node_ids.push(node_ids[nearest_node_indices[i]].clone());
+                }
+                Ok(nearest_node_ids)
+            },
+            Err(e) => Err(e),
+        }
+    }
+}
+
+impl PartialEq for NodeID {
+    fn eq(&self, nid: &NodeID) -> bool {
+        for i in 0..self.0.len() {
+            if self.0[i] != nid.0[i] {
+                return false;
+            }
+        }
+        (true)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use sha2::Sha256;
+    use tari_crypto::{
+        keys::SecretKey,
+        ristretto::{RistrettoPublicKey, RistrettoSecretKey},
+    };
+    use tari_utilities::byte_array::ByteArray;
+
+    #[test]
+    fn test_from_public_key() {
+        let mut rng = rand::OsRng::new().unwrap();
+        let sk = RistrettoSecretKey::random(&mut rng);
+        let pk = RistrettoPublicKey::from_secret_key(&sk);
+        let node_id = NodeID::from_public_key::<RistrettoPublicKey, Sha256>(&pk);
+        assert_ne!(node_id.0.to_vec(), NodeID::new().0.to_vec());
+        // Ensure node id is different to original public key
+        let mut pk_array: [u8; 32] = [0; 32];
+        pk_array.copy_from_slice(&pk.as_bytes());
+        assert_ne!(node_id.0.to_vec(), pk_array.to_vec());
+    }
+
+    #[test]
+    fn test_distance() {
```

Comment:
This isn't testing the `distance` function.

Reply:
Corrected, it should be test_node_id_ordering